### PR TITLE
9: Contains can now check network in network

### LINKED ISSUE #9: Contains function network in network
the `contains` function in networks now only accept Addresses, but it should also accept Networks.

### PR DESCRIPTION
**Pull request by noahlvb**
Resolves #9

### CODE REVIEW

**Comment by noahlvb** [comment]
I thought that crossed my mind was to add the method `public function address(): AddressInterface` to `RoutableInterface` but that would force me to implement `address()` on itself.

Also the name `RoutableInterface` is up for discussion but I couldn't find something that sat in between. Hope I will learn something from this :D

File: src/RoutableInterface.php [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Littledev\IPTools;
+
+interface RoutableInterface
```
Comment (blackshadev) [inline src/RoutableInterface.php:7]: If you add the address function maybe is `addressable` a better term, since we don't actually do routing and such.

**Review by blackshadev — CHANGES_REQUESTED** [review]
Some minor improvements.
Also you missed one implementation detail: 127.0.0.1/24 does not contain 127.0.0.1/8. The other way arround however is true.

An idea to implement this: Maybe add the prefix to the interface as wel and check on that as well in the contain function

File: src/RoutableInterface.php (hunk 1 above)
Comment (noahlvb) [reply src/RoutableInterface.php:7]: Good point changed it. I was thinking a network and a address are both routable but we don't do routing.

File: src/Network/IPv6Network.php [hunk 2]
```diff
@@ -45,8 +46,12 @@ public function subnet(): SubnetInterface
         return $this->subnet;
     }
 
-    public function contains(AddressInterface $address): bool
+    public function contains(RoutableInterface $address): bool
     {
+        if ($address instanceof NetworkInterface) {
+            $address = $address->address();
+        }
```
[the remark this answers is not in the document]
Comment (noahlvb) [reply src/Network/IPv6Network.php:53]: Sure was thinking the same but was not sure, will add it.

File: src/Network/IPv4Network.php [hunk 3]
```diff
@@ -42,10 +43,16 @@ public function subnet(): SubnetInterface
...
-    public function contains(AddressInterface $address): bool
+    public function contains(AddressableInterface $address): bool
     {
-        return (strcmp($address->inAddr(), $this->getFirstIP()->inAddr()) >= 0)
-            && (strcmp($address->inAddr(), $this->getLastIP()->inAddr()) <= 0);
+        if ($address->subnet() !== null
+            && !$this->subnet->contains($address->subnet())
+        ) {
+            return false;
+        }
```
Comment (blackshadev) [inline src/Network/IPv4Network.php:52]: I don't believe subnet needs to be nullable and with that the `$this->subnet->contains($address->subnet())` can be added to the end of the return statement. Which is nicer I think, it is also more logical to first check the address and after that the subnet, eventhough the result is the same it reads more logical

File: src/Network/IPv6Network.php [hunk 4]
```diff
@@ -45,10 +46,16 @@ public function subnet(): SubnetInterface
...
-    public function contains(AddressInterface $address): bool
+    public function contains(AddressableInterface $address): bool
     {
-        return (strcmp($address->inAddr(), $this->getFirstIP()->inAddr()) >= 0)
-            && (strcmp($address->inAddr(), $this->getLastIP()->inAddr()) <= 0);
+        if ($address->subnet() !== null
+            && !$this->subnet->contains($address->subnet())
+        ) {
+            return false;
+        }
```
Comment (blackshadev) [inline src/Network/IPv6Network.php:55]: Same as previous comment

File: src/Subnet/IPv4Subnet.php [hunk 5]
```diff
@@ -47,4 +46,9 @@ public function byteArray(): array
     {
         return $this->subnet->byteArray();
     }
+
+    public function contains(SubnetInterface $subnet): bool
+    {
+        return $subnet->prefix() >= $this->prefix();
```
Comment (blackshadev) [inline src/Subnet/IPv4Subnet.php:52]: Very nice, I would do this in the contains function of the address but this is way nicer. One extra check on the IP family is required in this case to make it 'correct'. With your implementation an IPv6/32 would contain an IPv4/32. Which isn't the case.

File: src/Subnet/IPv6Subnet.php [hunk 6]
```diff
@@ -47,4 +47,9 @@ public function __toString(): string
     {
         return (string)$this->subnet;
     }
+
+    public function contains(SubnetInterface $subnet): bool
+    {
+        return $subnet->prefix() >= $this->prefix();
```
Comment (blackshadev) [inline src/Subnet/IPv6Subnet.php:53]: See previous comment

File: tests/IPv4NetworkSpecTest.php [hunk 7]
```diff
@@ -54,6 +54,18 @@ public function testContainsWorksWithIPv6(): void
...
+    public function testContainsWorksWithNetworkInNetwork(): void
+    {
+        $network = IPv4Network::parse('127.0.0.1/16');
+
+        self::assertTrue($network->contains(IPv4Network::parse('127.0.0.1/16')));
+        self::assertTrue($network->contains(IPv4Network::parse('127.0.1.1/24')));
+        self::assertTrue($network->contains(IPv4Network::parse('127.0.0.4/23')));
+
+        self::assertFalse($network->contains(IPv4Network::parse('172.16.0.1/16')));
+        self::assertFalse($network->contains(IPv4Network::parse('127.0.0.1/8')));
```
Comment (blackshadev) [inline tests/IPv4NetworkSpecTest.php:66]: Could you add a test case for checking on IPv6 networks. An IPv4 network should never contain an IPv6 network

We can also track this in a separate issue if you want.

File: tests/IPv6NetworkSpecTest.php [hunk 8]
```diff
@@ -63,6 +63,18 @@ public function testContainsWorksWithIPv4(): void
...
+    {
+        $network = IPv6Network::parse('2001:db8::/64');
+
+        self::assertTrue($network->contains(IPv6Network::parse('2001:db8::/64')));
+        self::assertTrue($network->contains(IPv6Network::parse('2001:db8::42/122')));
+        self::assertTrue($network->contains(IPv6Network::parse('2001:db8::43/128')));
+//
+        self::assertFalse($network->contains(IPv6Network::parse('2001:db9::/64')));
+        self::assertFalse($network->contains(IPv6Network::parse('2001:db8::/48')));
+    }
```
Comment (blackshadev) [inline tests/IPv6NetworkSpecTest.php:76]: Could you add a test case for checking on IPv4 networks. An IPv6 network should never contain an IPv4 network

We can also track this in a separate issue if you want.

**Review by blackshadev — CHANGES_REQUESTED** [review]
Minor changes only now. Very nice work

File: src/AddressableInterface.php [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Littledev\IPTools;
+
+use Littledev\IPTools\Address\AddressInterface;
+use Littledev\IPTools\Subnet\SubnetInterface;
+
+interface AddressableInterface
+{
+    public function address(): AddressInterface;
+    public function subnet(): ?SubnetInterface;
```
[the remark this answers is not in the document]
Comment (noahlvb) [reply src/AddressableInterface.php:13]: That's left from before I figured that a address always has the subnet /32 or /128 so now both networks and address can always return a `SubnetInterface`

File: src/Network/IPv4Network.php (hunk 3 above)
Comment (noahlvb) [reply src/Network/IPv4Network.php:52]: Yea consider it done.

File: src/Subnet/IPv4Subnet.php (hunk 5 above)
Comment (noahlvb) [reply src/Subnet/IPv4Subnet.php:52]: Done

File: tests/IPv4NetworkSpecTest.php (hunk 7 above)
Comment (noahlvb) [reply tests/IPv4NetworkSpecTest.php:66]: Not creating another issue, doing it right the first time.

**Review by blackshadev — APPROVED** [review]

**Pull request merged**